packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,282 @@
+    const fileCache: FileCache = {}
+    const watchers = new Set<FSWatcher>()
+
+    // The "promise" event is for HMR updates to the pages map.
+    const emitPromise = debounce(() => {
+      this.emit('promise', pagesPromise)
+    }, 100)
+
+    // The "change" event is for HMR updates to each page.
+    const changedPages = new Set<string>()
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const setPageData = (pageId: string, pageData: Partial<PageData>) => {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `setPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      const key = pageData.key ?? 'main'
+      if (pageData.dataPath) {
+        if (page.data[key])
+          throw Error(
+            `setPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = pageData.dataPath
+      }
+      if (pageData.staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `setPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = pageData.staticData
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    const removePageData = (pageId: string, key?: string) => {
+      const page = pageCache[pageId]
+      if (page) {
+        if (key) {
+          delete page.data[key]
+          delete page.staticData[key]
+          changedPages.add(pageId)
+          emitChange()
+        } else {
+          delete pageCache[pageId]
+        }
+      }
+    }
+
+    const watchFiles = (
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) => {
+      let globs: string[]
+      let handler: (
+        file: File,
+        type: string
+      ) => PageData | Promise<PageData> | void
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = typeof arg2 === 'string' ? [arg2] : arg2!
+        handler = arg3 || defaultPageLoader
+      }
+
+      // Strip trailing slash and make absolute
+      baseDir = path.resolve(pagesDir, baseDir)
+
+      const processFile = async (
+        type: 'add' | 'change' | 'unlink',
+        filePath: string
+      ) => {
+        filePath = path.join(baseDir, filePath)
+
+        const file =
+          fileCache[filePath] ||
+          (fileCache[filePath] = new File(filePath, baseDir))
+
+        if (type === 'change') {
+          file.content = null
+          if (file.pageId) {
+            delete pageCache[file.pageId]
+          }
+        }
+
+        const result = await handler(file, type)
```

Comment:
If this is an 'unlink' event. It will try to `handle` a deleted file. Which may lead to error.